docs/design/interceptors.md

```diff
@@ -0,0 +1,691 @@
+ * serialized protocol specific request (e.g. HttpRequest), and the protocol specific response (e.g. HttpResponse).
+ */
+public interface ProtocolResponseInterceptorContext<I, ProtocolRequest, ProtocolResponse> :
+    ProtocolRequestInterceptorContext<I, ProtocolRequest> {
+    /**
+     * Retrieve the protocol specific response for the operation being invoked.
+     */
+    public val protocolResponse: ProtocolResponse
+}
+
+/**
+ * [Interceptor] context used for all phases that have access to the operation input (request), the
+ * serialized protocol specific request (e.g. HttpRequest), the protocol specific response (e.g. HttpResponse),
+ * and the deserialized operation output (response).
+ */
+public interface ResponseInterceptorContext<I, O, ProtocolRequest, ProtocolResponse> :
+    ProtocolResponseInterceptorContext<I, ProtocolRequest, ProtocolResponse> {
+    /**
+     * Retrieve the modeled response or exception for the operation being invoked
+     */
+    public val response: Result<O>
+}
+
+```
+
+The `Interceptor` interface defined above is generic over:
+* `Input` - the operation input type
+* `Output` - the operation output type
+* `ProtocolRequest` - the specific protocol (transmit) request type
+* `ProtocolResponse` - the specific protocol (transmit) response type
+
+
+Observations:
+
+* Every interceptor method has a default implementation. This allows customers to override only the hooks they are 
+interested in as well as allowing future hooks to be defined without breaking API or binary compatibility.
+* Multiple context types are defined that restrict the set of information available to a particular hook (e.g. you can't
+access the operation'output in any hook until *after* the first deserialization hook). Additionally the hooks are 
+specific about the nullability of particular context information (e.g. the `modifyBeforeCompletion` hook may or may not
+have a protocol request or response available depending on how far the operation execution made it.
+* Hooks are non-blocking (note no `suspend` in any hook method) and implementators are expected to respect this
+* Despite being generic over the input and output type, the `Input` and `Output` types of an `Interceptor` are expected
+to be `Any` in practice. This is because interceptors can be registered at the *client* level. Such an interceptor would
+be executed for every operation and would not be able to know the specific input or output type at compile time. 
+Interceptors that execute for any operation are common and expected (e.g. add a header to every outgoing request). 
+The downside to this flexibility is that customers will have to cast the input or output type from the context to the 
+specific operation type. 
+  * NOTE: The modeled operation name will be available in the context (it is already available in the execution context)
+  which can be used on interceptors registered at the client level that want to deal with specific operations.
```

Comment:
**Comment**: In principle, what do we think about providing types for hooks which only execute on a single operation (and ignore other operations)? The benefit would be allowing users to easily add strongly-typed interceptors without casting or checking themselves:

```kotlin
val client = S3Client {
  interceptors += ListObjectsMaxKeysInterceptor(10)
}

class ListObjectsMaxKeysInterceptor(defaultMaxKeys: Int): ListObjectsInterceptor() {
    override fun modifyBeforeSerialization(context: RequestInterceptorContext<ListObjectsRequest>) = when {
        context.request.maxKeys == null -> context.request.copy { maxKeys = defaultMaxKeys }
        else -> context.request
    }
}
```

To achieve this, I think we'd need the following:

### Remove input/output types on `Interceptor`

```kotlin
interface Interceptor<ProtocolRequest, ProtocolResponse> {
    fun modifyBeforeSerialization(context: RequestInterceptorContext<Any>): Any = context.request
    fun modifyBeforeCompletion(context: ResponseInterceptorContext<Any, Any, ProtocolRequest?, ProtocolResponse?>): Result<Any> = context.response
}

typealias HttpInterceptor = Interceptor<HttpRequest, HttpResponse>
```

(only including a few hooks for demonstration purposes)

### Add a new `TypedInterceptor` with input/output types (effectively the old `Interceptor`)

```kotlin
interface TypedInterceptor<Input, Output, ProtocolRequest, ProtocolResponse> {
    fun modifyBeforeSerialization(context: RequestInterceptorContext<Input>): Input = context.request
    fun modifyBeforeCompletion(context: ResponseInterceptorContext<Input, Output, ProtocolRequest?, ProtocolResponse?>): Result<Output> = context.response
}

typealias TypedHttpInterceptor<Input, Output> = TypedInterceptor<Input, Output, HttpRequest, HttpResponse>
```

### Codegen per-op interceptor classes

Example for **ListObjects**:

```kotlin
open class ListObjectsInterceptor : HttpInterceptor, TypedHttpInterceptor<ListObjectsRequest, ListObjectsResponse> {
    final override fun modifyBeforeSerialization(context: RequestInterceptorContext<Any>): Any = when {
        context.opName == "ListObjects" -> {
            val typedContext = object : RequestInterceptorContext<ListObjectsRequest> {
                override val opName = "ListObjects"
                override val request = context.request as ListObjectsRequest
            }
            modifyBeforeSerialization(typedContext)
        }
        else -> context.request
    }
    
    final override fun modifyBeforeCompletion(context: ResponseInterceptorContext<Any, Any, HttpRequest?, HttpResponse?>): Result<Any> = when {
        context.opName == "ListObjects" -> {
            val typedContext = object : ResponseInterceptorContext<ListObjectsRequest, ListObjectsResponse, HttpRequest?, HttpResponse?> {
                override val opName = "ListObjects"
                override val request = context.request as ListObjectsRequest
                override val protocolRequest = context.protocolRequest
                override val protocolResponse = context.protocolResponse
                override val response = context.response as Result<ListObjectsResponse>
            }
            modifyBeforeCompletion(typedContext)
        }
        else -> context.response
    }
}
```

Ignore the inline anonymous contexts, I'm assuming we'll have some default implementation class to use at runtime. Also, the operation name checking and casting may be reusable enough to encapsulate in a base class to shrink the per-op interceptor classes down to just the types and operation name...not sure yet.

Reply:
Ya something like this may work and is in back of my mind. I actually think we _may_ not even need new types depending on if we thread `Any` as a bound on some of the generics. 

Let's get interceptors wired up and before we land it come back to revisit typed interceptors (it will give us more room to play with these ideas once we have a working framework).